File: JAVA/Data Structure/Singly_Circular_LL.java
import java.util.Scanner ;

class Node
{
    int data ;
    Node next ;

    public Node(int iNo)
    {
        data = iNo ;
        next = null ;
    }
}

class SinglyCircular
{
    // characterstics
    public Node head ;
    public Node tail ;
    public int iCountNode ;

    // constructor
    public SinglyCircular()
    {
        head = null ;
        tail = null ;
        iCountNode = 0 ;
    }

    // countNode method is used for count the peresent elements in LinkedList.
    public int countNode()
    {
        if(head == null)
        {
            return 0 ;
        }

        return iCountNode ;
    }

    // displayLL method is used for display the elements which are present inside the LinkedList
    public void displayLL()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is empty");
            return ;
        }

        Node temp = head ;

        System.out.print("<=>");
        do
        {
            System.out.print("|" + temp.data + " |<=>");
            temp = temp.next ;
        }
        while(temp != tail.next);
        System.out.println();
    }

    // insertFirst method is used for insert the new node inside the LinkedList at first position
    public void insertFirst(int iNo)
    {
        Node newn = null ;

        newn = new Node(iNo);

        if(head == null && tail == null)
        {
            head = newn ;
            tail = newn ;
        }
        else
        {
            newn.next = head ;
            head = newn ;
        }
        tail.next = head ;

        iCountNode++ ;

        System.out.println("Data inserted succesfully.");
    }

    // insertLast method is used for insert the new node inside the LinkedList at Last position
    public void insertLast(int iNo)
    {
        Node newn = null ;

        newn = new Node(iNo) ;

        if(head == null && tail == null)
        {
            head = newn ;
            tail = newn ;
        }
        else
        {
            tail.next = newn ;
            tail = newn ;
        }
        tail.next = head ;
        iCountNode++ ;
        System.out.println("Data inserted succesfully.");
    }

    // insertAtPos method is used for insert new node any your required position in LL 
    public void insertAtPos(int iNo , int iPos)
    {
        if(iPos < 1 || iPos > iCountNode+1)
        {
            System.out.println("Invalid Position !! , Please enter correct position.");
            return ;
        }

        if(iPos == 1)
        {
            insertFirst(iNo);
        }
        else if(iPos == iCountNode+1)
        {
            insertLast(iNo);
        }
        else
        {
            Node newn = null ;
            int i = 0 ;

            Node temp = head ;

            newn = new Node(iNo);

            for(i = 1 ; i < iPos-1 ; i++)
            {
                temp = temp.next ;
            }
            newn.next = temp.next ;
            temp.next = newn ;

            iCountNode++ ;

            System.out.println("Data inserted succesfully.");
        }
    }

    // deleteFirst method is used to Delete the first node of the LinkedList
    public void deleteFirst()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is allready empty , Unable to delete");
            return ;
        }

        if(head == tail)
        {
            head = null ;
            tail = null ;
        }
        else
        {
            head = head.next ;

            tail.next = head ;
        }
        iCountNode-- ;

        System.out.println("Data deleted succesfully.");
    }

    // deleteLast method is used to Delete the last node of the LinkedList
    public void deleteLast()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is allready empty , Unable to delete");
            return ;
        }

        if(head == tail)
        {
            head = null ;
            tail = null ;
        }
        else
        {
            Node temp = head ;

            while(temp.next != tail)
            {
                temp = temp.next ;
            }
            tail = temp ;
            tail.next = head ;
        }
        iCountNode-- ;

        System.out.println("Data deleted succesfully.");
    }

    // deleteAtPos is used to delete the node from required position of LL 
    public void deleteAtPos(int iPos)
    {
        if(iPos < 1 || iPos > iCountNode)
        {
            System.out.println("Invalid Position !!");
            return ;
        }

        if(iPos == 1)
        {
            deleteFirst();
        }
        else if(iPos == iCountNode)
        {
            deleteLast();
        }
        else
        {
            Node temp = head ;
            int i = 0 ;

            for(i=1; i < iPos-1 ; i++)
            {
                temp = temp.next ;
            }
            temp.next = temp.next.next ;

            iCountNode-- ;
            
            System.out.println("Data deleted succesfully.");
        }
        tail.next = head ;
    }

    // emptyLL method is used to delete all nodes and make LL empty
    public void emptyLL()
    {
        if(head == null)
        {
            System.out.println("LL is already empty , Unable to use emptyLL method");
            return ;
        }

        Node nextNode = null ;
        Node temp = head ;

        for(int i = 1 ; i <= iCountNode ; i++)
        {
            nextNode = temp.next ;
            temp.next = null ;
            temp = nextNode ;
        }

        head = null ;
        tail = null ;

        iCountNode = 0 ;

        System.out.println("Now , LL is empty");
    }
}

class Singly_Circular_LL
{
    public static void main(String arg[])
    {
        int iOption = 0 , iRet = 0 , iData = 0 , iPos = 0 ;

        SinglyCircular sobj = new SinglyCircular();

        Scanner scobj = new Scanner(System.in);

        System.out.println("NOTE :\n1 : position of LL start with 1");

        while(true)
        {
            System.out.println("-----------------------------------------------------------------");
            System.out.println("0 : Terminate the Application");
            System.out.println("1 : Display the LL ");
            System.out.println("2 : Count the total nodes of LL");
            System.out.println("3 : Insert the data at first position");
            System.out.println("4 : Insert the data at last position");
            System.out.println("5 : Insert the data at required position");
            System.out.println("6 : Delete the data of first position");
            System.out.println("7 : Delete the data of last position");
            System.out.println("8 : Delete the data from requrired position");
            System.out.println("9 : Delete the all data from LL (OR) make LL empty");
            System.out.println("-----------------------------------------------------------------");
            System.out.println("Enter the option : ");
            iOption = scobj.nextInt();
            System.out.println("-----------------------------------------------------------------");

            switch(iOption)
            {
                case 0 :
                    System.out.println("Application is closed..");
                    scobj.close();
                    return ;

                case 1 :
                    sobj.displayLL();
                    break ;

                case 2 :
                    iRet = sobj.countNode(); 
                    System.out.println("Total nodes in LL are : "+iRet);
                    break ;

                case 3 : 
                    System.out.println("Enter the data : ");
                    iData = scobj.nextInt();

                    sobj.insertFirst(iData);
                    break ;

                case 4 :
                    System.out.println("Enter the data : ");
                    iData = scobj.nextInt();

                    sobj.insertLast(iData);
                    break ;

                case 5 :
                    System.out.println("Enter the data : ");
                    iData = scobj.nextInt();

                    System.out.println("Enter the position : ");
                    iPos = scobj.nextInt();

                    sobj.insertAtPos(iData,iPos);
                    break ;

                case 6 :
                    sobj.deleteFirst();
                    break ;

                case 7 :
                    sobj.deleteLast();
                    break ;

                case 8 :
                    System.out.println("Enter the position : ");
                    iPos = scobj.nextInt();

                    sobj.deleteAtPos(iPos);
                    break ;

                case 9 :
                    sobj.emptyLL();
                    break ;

                default :
                    System.out.println("!! Invalid Option !!");
                    break ;          
            }
        }
    }
}

File: C++/Data Structure/Singly_Circular_LL.cpp
#include<iostream>
using namespace std;

struct node 
{
    int data ;
    struct node* next ;
};

typedef struct node NODE ;
typedef struct node* PNODE;

class Singly_Circular
{
    public:
        PNODE head ;
        PNODE tail ;
        int iCountNode ;

        Singly_Circular()
        {
            head = NULL ;
            tail = NULL ;
            iCountNode = 0 ;
        }

        int CountNode()
        {
            return iCountNode ;
        }

        void Display()
        {
            if(head == NULL || tail == NULL)
            {
                cout<<"LL is empty !!\n";
                return ;
            }

            PNODE temp = NULL ;
            temp = head ;

            do
            {
                cout<<"|"<<temp->data<<"|->";
                temp = temp->next ;
            }
            while(temp != (tail->next));

            cout<<"NULL\n";
        }

        void InsertFirst(int iNo)
        {
            PNODE newn = NULL ;

            newn = new NODE ;

            newn->data = iNo ;
            newn->next = NULL ;

            if(head == NULL && tail == NULL)
            {
                head = newn ;
                tail = newn ;
            }
            else
            {
                newn->next = head ;
                head = newn ;
            }
            tail->next = head ;
            iCountNode++ ;
        }

        void InsertLast(int iNo)
        {
            PNODE newn = NULL ;

            newn = new NODE ;

            newn->data = iNo ;
            newn->next = NULL ;

            if(head == NULL && tail == NULL)
            {
                head = newn ;
                tail = newn ;
            }
            else
            {
                tail->next = newn ;
                tail = newn ;
            }
            tail->next = head ;
            iCountNode++ ;
        }

        void DeleteFirst()
        {
            if(head == NULL && tail == NULL)
            {
                cout<<"LL is already empty\n";
                return ;
            }
            else if(head == tail)
            {
                delete head;

                head = NULL ;
                tail = NULL ;
            }
            else
            {
                head = head->next ;
                delete (tail->next);

                tail->next = head ;
            }    
            iCountNode-- ;
        }

        void DeleteLast()
        {
            if(head == NULL && tail == NULL)
            {
                cout<<"LL is already empty\n";
                return ;
            }
            else if(head == tail)
            {
                delete head;

                head = NULL ;
                tail = NULL ;
            }
            else
            {
                PNODE temp = NULL ;
                PNODE target = NULL ;

                temp = head ;

                while(temp->next != tail)
                {
                    temp = temp->next ;
                }
                delete (temp->next);
                tail = temp ;

                tail->next = head ;
            }
            iCountNode-- ;
        }

        void InsertAtPos(int iNo , int iPos)
        {
            if(iPos < 1 || iPos > iCountNode+1)
            {
                cout<<"Invalid Position !!\n";
                return ;
            }
            else if(iPos == 1)
            {
                InsertFirst(iNo);
            }
            else if(iPos == iCountNode+1)
            {
                InsertLast(iNo);
            }
            else
            {
                PNODE temp = NULL ;
                PNODE newn = NULL ;
                int i = 0 ;

                newn = new NODE ;

                newn->data = iNo ;
                newn->next = NULL ;

                temp = head ;

                for(i = 1 ; i < iPos-1 ; i++)
                {
                    temp = temp->next ;
                }
                newn->next = temp->next ;
                temp->next = newn ;

                iCountNode++ ;
            }
        }

        void DeleteAtPos(int iPos)
        {
            if(iPos < 1 || iPos > iCountNode)
            {
                cout<<"Invalid Position !!\n";
                return ;
            }
            else if(iPos == 1)
            {
                DeleteFirst();
            }
            else if(iPos == iCountNode)
            {
                DeleteLast();
            }
            else
            {
                PNODE temp = NULL ;
                PNODE target = NULL ;
                int i = 0 ;

                temp = head ;

                for(i = 1 ; i < iPos-1 ; i++)
                {
                    temp = temp->next ;
                }
                target = temp->next ;
                temp->next = temp->next->next ;
                delete (target) ;

                iCountNode-- ;
            }
        }

        void DeleteAllNode()
        {
            if(head == NULL) return;

            PNODE temp = head;
            PNODE next = NULL;

            do
            {
                next = temp->next;
                delete temp;
                temp = next;
            }
            while(temp != head);

            head = NULL;
            tail = NULL;
            iCountNode = 0;
        }     
};

int main()
{
    int iOption = 0 , iNo = 0 , iPos = 0 , iRet = 0  ;

    Singly_Circular sobj ;

    while(1)
    {
        cout<<"_________________________________________________________________________\n";
        cout<<"Singly Circular Linked List started...\n";
        cout<<"_________________________________________________________________________\n";
        cout<<"1 Insert Element at first position\n";
        cout<<"2 Insert Element at Last position\n";
        cout<<"3 Insert Element at Required position, (position start with 1)\n";
        cout<<"4 Delete first node \n";
        cout<<"5 Delete last node \n";
        cout<<"6 Delete node from Required position, (position start with 1)\n";
        cout<<"7 Display Linked List\n";
        cout<<"8 Count of Total elemnets present in Linked List\n";
        cout<<"9 Delete all elements from the Linked List\n";
        cout<<"10 Terminate the Application\n";
        cout<<"_________________________________________________________________________\n";
        cout<<"Enter you option : \n";
        cin>>iOption;
        cout<<"_________________________________________________________________________\n";

        if(iOption == 1)
        {
            cout<<"Enter the data : \n";
            cin>>iNo ;

            sobj.InsertFirst(iNo);
        }
        else if(iOption == 2)
        {
            cout<<"Enter the data : \n";
            cin>>iNo ;

            sobj.InsertLast(iNo);
        }
        else if(iOption == 3)
        {
            cout<<"Enter the data : \n";
            cin>>iNo ;

            cout<<"Enter the position : \n";
            cin>>iPos ;

            sobj.InsertAtPos(iNo,iPos);
        }
        else if(iOption == 4)
        {
            sobj.DeleteFirst();
        }
        else if(iOption == 5)
        {
            sobj.DeleteLast();
        }
        else if(iOption == 6)
        {
            cout<<"Enter the position : \n";
            cin>>iPos ;

            sobj.DeleteAtPos(iPos);
        }
        else if(iOption == 7)
        {
            sobj.Display();
        }
        else if(iOption == 8)
        {
            iRet = sobj.CountNode();
            cout<<"Total nodes in LL are : "<<iRet<<"\n";
        }
        else if(iOption == 9)
        {
            sobj.DeleteAllNode();
        }
        else if(iOption == 10)
        {
            cout<<"Application is terminated.\n";
            break;
        }
        else
        {
            cout<<"Invalid option !!! \n";
        }
    }

    return 0 ;
}